### ISSUE TITLE
[question] usage of options in set_version

### ISSUE THREAD

**Issue by hubertba** [conan-io/conan#15208]
### What is your question?

Hi

I would like to create packages with and with our a timestamp in the version number.
So I try to use an option called with_timestamp and read that option in the set_version function.
But this does not work as expected.

this is my recipe:

```
import os
import pdb

from conan import ConanFile
from conan.tools.files import load
from datetime import datetime, timezone

class ComponentConanFile(ConanFile):
    name = "demo"
    version = "1.2.3"    
    options = {
        "with_timestamp" : [True, False]  
        }

    def config_options(self):   
        if self.options.with_timestamp == None:
            self.options.with_timestamp = True
    
    def set_version(self):
        #breakpoint()
        orig_version = self.version
        self.output.info("ORIG: " + orig_version)
        try:
            option_dict = vars(self.options.with_timestamp)
            print("Content of self.options.with_timestamp:")
            print(option_dict)
            if self.options.with_timestamp:
                current_datetime = datetime.utcnow().replace(tzinfo=timezone.utc)
                formatted_datetime = current_datetime.strftime("%Y%m%d-%H%M")
                print(formatted_datetime)
                self.version = orig_version + "-" + formatted_datetime
                print(self.version)
        except Exception as e:
            self.output.error(e)

```

Is it the intended behaviour that options are not available within set_version? 

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @hubertba 

Thanks for your question.

Yes, it is definitely the intended behavior. ``options`` are a parametrization of the "binaries" that can be created for a give version, but not a parametrization for the version itself. Same as ``settings``, both are modeling the binaries, not the version.

What you can probably do, instead of trying to use a conf for this, is using the ``--version=xxxx`` argument directly? Then, you can compose that input with other source at your will within the ``set_version()`` method to create the final version number?

**Comment by hubertba** [comment]
not really ... I tried it already ... and I get following error:
```

class ComponentConanFile(ConanFile):
    name = "demo"
  
    def set_version(self):
        self.version = "1.2.3"


➜  conan2_component git:(main) ✗ conan create . --version="1.2.3-20231204-1205"

======== Exporting recipe to the cache ========
ERROR: Package recipe with version 1.2.3-20231204-1205!=1.2.3
```
and I would love to prevent that I need to give there version number at each and every command.

is there any possibility to access any parameter value given from the CLI in the set_version command?

**Comment by memsharded** [comment]
> is there any possibility to access any parameter value given from the CLI in the set_version command?

Yes, you have it available in the ``self.version`` field.

Note the examples in https://docs.conan.io/2/reference/conanfile/methods/set_version.html, you can use the  existing ``self.version`` value from CLI and compose it with your other values, use conditionals, etc. See the example of ``self.version = self.version or load(self, "version.txt")``

**Comment by hubertba** [comment]
Thank you very much ... the conditional is working ... now I just have to find out how to handle it from the workflow perspective.

**Comment by memsharded** [comment]
You might "abuse" a bit the ``--version=with_timestamp`` and read that exact value for doing what you want.

In any case, it would be interesting to know the root need for this. Why are you adding timestamps to the versions? How are you adding that timestamp, like a ``+build.timestamp`` metadata? How are you consuming that version later, via version-ranges? What problems are you trying to solve with that timestamp?  

That might be good feedback, maybe we can also advice some possible approaches. Feel free to open a new ticket to discuss it, this seems good closed as the ``set_version()`` approach seems to address this need well enough.

**Comment by hubertba** [comment]
which still gives me some headache

**Comment by hubertba** [comment]
in my 
```
def set_version(self):
        try:
            #breakpoint()
            content = load(self, "_metadata/mic/types.mic")
            versionff = re.search("deliveryVersion = (.*)", content).group(1)
            versionff.strip()
            if ( self.version =="timestamp"):
                current_datetime = datetime.utcnow().replace(tzinfo=timezone.utc)
                formatted_datetime = current_datetime.strftime("%Y%m%d-%H%M")
                print(formatted_datetime)
                self.version = versionff + '-' + formatted_datetime
            self.version = versionff
            print(self.version)
        except Exception as e:
            self.output.error(e)
```
But if I run it I get:
20231204-1602
0.0.2
ERROR: Package recipe with version timestamp!=0.0.2

**Comment by memsharded** [comment]
You are right, I overlooked one aspect, is that the version can be defined if not yet, but cannot be modified. I am having a look at it, re-opening in the meantime.

**Comment by memsharded** [comment]
I am exploring in https://github.com/conan-io/conan/pull/15211, if we can relax the checks on ``set_version()`` so your flow would be allowed.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/15211 has been merged, it will be in next 2.0.15 release, allowing the above behavior in ``set_xxx()`` methods
